### ISSUE TITLE
Segmentation fault when deleting files that are bigger than 4MB

### ISSUE THREAD

**Issue by JerkFace2017** [zfsonlinux/zfs#8832]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  Fedora
Distribution Version	| 30
Linux Kernel	| 5.0.17-300
Architecture	| x86
ZFS Version	| 0.8.0-1
SPL Version	| 0.8.0-1

### Describe the problem you're observing
When trying to delete a file larger than 4MB in size the command returns a Segmentation fault. I have recently upgraded to zfs 0.8.0-1 on my Fedora 30 box. I went to clean up some old files (~500 gig) that were being shared via a samba share and the process was taking a unusually long time to remove the files. I logged into the box to remove the files from the filesystem and to see what was taking so long. When I entered rm -Rf <dir name>I got a Segmentation fault. At first glance it looked like noting was deleted in the directory. After a closer look the first file selected by rm was removed but then the command Seg faults and I get a core dump/huge error in the log file.

After a few days of messing around I discovered that files and directories under 3 MB in size worked as expected but anything over 4 MB gave a Seg fault.

I pulled out a spare HD and created a default zpool  with a default file system on it for testing. Everything worked fine for any file size or directory. I compared the parameters and remembered that I set the record size to 4M a while back for some testing. Soon as i set the recordsize below 4M everything started working again .

### Describe how to reproduce the problem
zfs set recordsize=4M pool/filesystem

### Include any warning/errors/backtraces from the system logs

Terminal output and strace output when recordsize is set to 4M and deleting a 100 MB file called file.txt

```
[root@deadpool testing]# ls -l
total 102264
-rw-r--r-- 1 root root 104857600 May 29 21:27 file.txt
[root@deadpool testing]# du file.txt
102264  file.txt
[root@deadpool testing]# rm file.txt
rm: remove regular file 'file.txt'? y
Segmentation fault
[root@deadpool testing]#
[root@deadpool testing]# zfs get recordsize deadpool
NAME      PROPERTY    VALUE    SOURCE
deadpool  recordsize  4M       local
[root@deadpool testing]#

execve("/usr/bin/rm", ["rm", "file.txt"], 0x7ffc4375f418 /* 23 vars */) = 0
brk(NULL)                               = 0x55917d590000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffdd05183c0) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=164321, ...}) = 0
mmap(NULL, 164321, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fd956891000
close(3)                                = 0
openat(AT_FDCWD, "/lib64/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@@\2\0\0\0\0\0"..., 832) = 832
lseek(3, 792, SEEK_SET)                 = 792
read(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\2605Q\367\26\207U\27\277\347\342M\3541\24{"..., 68) = 68
fstat(3, {st_mode=S_IFREG|0755, st_size=5591560, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd95688f000
lseek(3, 792, SEEK_SET)                 = 792
read(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\2605Q\367\26\207U\27\277\347\342M\3541\24{"..., 68) = 68
lseek(3, 864, SEEK_SET)                 = 864
read(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32) = 32
mmap(NULL, 1857568, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fd9566c9000
mprotect(0x7fd9566eb000, 1679360, PROT_NONE) = 0
mmap(0x7fd9566eb000, 1363968, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7fd9566eb000
mmap(0x7fd956838000, 311296, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x16f000) = 0x7fd956838000
mmap(0x7fd956885000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bb000) = 0x7fd956885000
mmap(0x7fd95688b000, 14368, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fd95688b000
close(3)                                = 0
arch_prctl(ARCH_SET_FS, 0x7fd956890580) = 0
mprotect(0x7fd956885000, 16384, PROT_READ) = 0
mprotect(0x55917d07f000, 4096, PROT_READ) = 0
mprotect(0x7fd9568e4000, 4096, PROT_READ) = 0
munmap(0x7fd956891000, 164321)          = 0
brk(NULL)                               = 0x55917d590000
brk(0x55917d5b1000)                     = 0x55917d5b1000
brk(NULL)                               = 0x55917d5b1000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=217750512, ...}) = 0
mmap(NULL, 217750512, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fd94971f000
close(3)                                = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
newfstatat(AT_FDCWD, "file.txt", {st_mode=S_IFREG|0644, st_size=104857600, ...}, AT_SYMLINK_NOFOLLOW) = 0
geteuid()                               = 0
unlinkat(AT_FDCWD, "file.txt", 0)       = ?
+++ killed by SIGSEGV +++

```
Here is what's in /var/log/messages when the rm returns a Segmentation fault:
```
May 29 21:33:49 deadpool kernel: divide error: 0000 [#16] SMP PTI
May 29 21:33:49 deadpool kernel: CPU: 3 PID: 5087 Comm: rm Tainted: P      D    OE     5.0.17-300.fc30.x86_64 #1
May 29 21:33:49 deadpool kernel: Hardware name: Gigabyte Technology Co., Ltd. H97-D3H/H97-D3H-CF, BIOS F5 06/26/2014
May 29 21:33:49 deadpool kernel: RIP: 0010:dmu_free_long_range+0x153/0x500 [zfs]
May 29 21:33:49 deadpool kernel: Code: 41 bd 00 00 00 04 31 d2 4c 8d 14 03 41 0f b6 46 6b 4c 89 54 24 60 8d 48 01 41 d3 fd 8d 48 f9 41 d3 e4 4d 63 ed 4b 8d 44 22 ff <49> f7 f4 31 d2 49 89 c7 48 89 d8 49 f7 f4 31 d2 49 29 c7 4c 89 f8
May 29 21:33:49 deadpool kernel: RSP: 0018:ffffab472a33fd48 EFLAGS: 00010247
May 29 21:33:49 deadpool kernel: RAX: 00000000063fffff RBX: 0000000000000000 RCX: 000000000000000a
May 29 21:33:49 deadpool kernel: RDX: 0000000000000000 RSI: 00000000fac32bff RDI: ffff8a8a481b8ce8
May 29 21:33:49 deadpool kernel: RBP: ffff8a8a72419000 R08: 0000000000000000 R09: 0000000000000000
May 29 21:33:49 deadpool kernel: R10: 0000000006400000 R11: 7478742e656c6966 R12: 0000000000000000
May 29 21:33:49 deadpool kernel: R13: 0000000000000100 R14: ffff8a89b1836300 R15: ffff8a8a481b8800
May 29 21:33:49 deadpool kernel: FS:  00007f23d9a47580(0000) GS:ffff8a8a7db80000(0000) knlGS:0000000000000000
May 29 21:33:49 deadpool kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
May 29 21:33:49 deadpool kernel: CR2: 000055b5a40ee048 CR3: 000000073d292005 CR4: 00000000001606e0
May 29 21:33:49 deadpool kernel: Call Trace:
May 29 21:33:49 deadpool kernel: ? mutex_lock+0xe/0x30
May 29 21:33:49 deadpool kernel: ? zfs_znode_hold_enter+0x111/0x170 [zfs]
May 29 21:33:49 deadpool kernel: zfs_rmnode+0x256/0x330 [zfs]
May 29 21:33:49 deadpool kernel: ? zfs_zinactive+0xd1/0xf0 [zfs]
May 29 21:33:49 deadpool kernel: zfs_inactive+0x82/0x200 [zfs]
May 29 21:33:49 deadpool kernel: ? unmap_mapping_pages+0x5e/0x130
May 29 21:33:49 deadpool kernel: zpl_evict_inode+0x3e/0x50 [zfs]
May 29 21:33:49 deadpool kernel: evict+0xcb/0x190
May 29 21:33:49 deadpool kernel: do_unlinkat+0x1c8/0x2e0
May 29 21:33:49 deadpool kernel: do_syscall_64+0x5b/0x170
May 29 21:33:49 deadpool kernel: entry_SYSCALL_64_after_hwframe+0x44/0xa9
May 29 21:33:49 deadpool kernel: RIP: 0033:0x7f23d996daeb
May 29 21:33:49 deadpool kernel: Code: 73 01 c3 48 8b 0d 9d 23 0d 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 90 f3 0f 1e fa b8 07 01 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 6d 23 0d 00 f7 d8 64 89 01 48
May 29 21:33:49 deadpool kernel: RSP: 002b:00007fff665a0948 EFLAGS: 00000246 ORIG_RAX: 0000000000000107
May 29 21:33:49 deadpool kernel: RAX: ffffffffffffffda RBX: 000055b5a40eb6c0 RCX: 00007f23d996daeb
May 29 21:33:49 deadpool kernel: RDX: 0000000000000000 RSI: 000055b5a40ea490 RDI: 00000000ffffff9c
May 29 21:33:49 deadpool kernel: RBP: 000055b5a40ea400 R08: 000055b5a40ec040 R09: 0000000000000000
May 29 21:33:49 deadpool kernel: R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000000
May 29 21:33:49 deadpool kernel: R13: 00007fff665a0b20 R14: 00007fff665a0b20 R15: 0000000000000002
May 29 21:33:49 deadpool kernel: Modules linked in: fuse uas usb_storage nf_conntrack_netbios_ns nf_conntrack_broadcast xt_CT ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 xt_conntrack ebtable_nat ip6table_nat nf_nat_ipv6 ip6table_mangle ip6table_raw ip6table_security iptable_nat nf_nat_ipv4 nf_nat iptable_mangle iptable_raw iptable_security nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c ip_set nfnetlink ebtable_filter ebtables ip6table_filter ip6_tables zfs(POE) intel_rapl zunicode(POE) zavl(POE) icp(POE) x86_pkg_temp_thermal intel_powerclamp zlua(POE) coretemp kvm_intel mei_wdt kvm iTCO_wdt iTCO_vendor_support zcommon(POE) znvpair(POE) spl(OE) irqbypass crct10dif_pclmul crc32_pclmul snd_soc_rt5640 snd_soc_rl6231 ghash_clmulni_intel intel_cstate snd_hda_codec_realtek intel_uncore snd_hda_codec_generic ledtrig_audio intel_rapl_perf snd_soc_core snd_hda_intel snd_hda_codec snd_compress ac97_bus snd_pcm_dmaengine mei_me snd_seq mei snd_hda_core lpc_ich i2c_i801 snd_hwdep snd_seq_device snd_pcm snd_timer snd
May 29 21:33:49 deadpool kernel: soundcore acpi_pad pcc_cpufreq binfmt_misc nouveau mxm_wmi wmi i2c_algo_bit drm_kms_helper crc32c_intel ttm serio_raw drm e1000e video
May 29 21:33:49 deadpool kernel: ---[ end trace 01e54ee9fdc5c1f0 ]---
May 29 21:33:49 deadpool kernel: RIP: 0010:dmu_free_long_range+0x153/0x500 [zfs]
May 29 21:33:49 deadpool kernel: Code: 41 bd 00 00 00 04 31 d2 4c 8d 14 03 41 0f b6 46 6b 4c 89 54 24 60 8d 48 01 41 d3 fd 8d 48 f9 41 d3 e4 4d 63 ed 4b 8d 44 22 ff <49> f7 f4 31 d2 49 89 c7 48 89 d8 49 f7 f4 31 d2 49 29 c7 4c 89 f8
May 29 21:33:49 deadpool kernel: RSP: 0018:ffffab470b9b3ab0 EFLAGS: 00010247
May 29 21:33:49 deadpool kernel: RAX: 00000000063fffff RBX: 0000000000000000 RCX: 000000000000000a
May 29 21:33:49 deadpool kernel: RDX: 0000000000000000 RSI: 00000000fac32bff RDI: ffff8a8a481b8ce8
May 29 21:33:49 deadpool kernel: RBP: ffff8a8a72419000 R08: 0000000000000001 R09: 0000000000000000
May 29 21:33:49 deadpool kernel: R10: 0000000006400000 R11: 0000000000000008 R12: 0000000000000000
May 29 21:33:49 deadpool kernel: R13: 0000000000000100 R14: ffff8a8a2bfc1880 R15: ffff8a8a481b8800
May 29 21:33:49 deadpool kernel: FS:  00007f23d9a47580(0000) GS:ffff8a8a7db80000(0000) knlGS:0000000000000000
May 29 21:33:49 deadpool kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
May 29 21:33:49 deadpool kernel: CR2: 000055b5a40ee048 CR3: 000000073d292005 CR4: 00000000001606e0
May 29 21:33:50 deadpool abrt-dump-journal-oops[1303]: abrt-dump-journal-oops: Found oopses: 1
May 29 21:33:50 deadpool abrt-dump-journal-oops[1303]: abrt-dump-journal-oops: Creating problem directories
May 29 21:33:50 deadpool abrt-server[5088]: Deleting problem directory oops-2019-05-29-21:33:50-1303-0 (dup of oops-2019-05-24-11:43:07-2811-0)
May 29 21:33:50 deadpool abrt-notification[5095]: System encountered a non-fatal error in mutex_lock()
May 29 21:33:51 deadpool abrt-dump-journal-oops[1303]: Reported 1 kernel oopses to Abrt
```
Change the record size to 2M and everything works as expected:
```
[root@deadpool tank]# zfs set recordsize=2M deadpool
[root@deadpool tank]# zfs get recordsize deadpool
NAME      PROPERTY    VALUE    SOURCE
deadpool  recordsize  2M       local
[root@deadpool tank]#

[root@deadpool testing]# ls -l
total 102264
-rw-r--r-- 1 root root 104857600 May 29 21:35 file.txt
[root@deadpool testing]# du file.txt
102264  file.txt
[root@deadpool testing]# zfs get recordsize deadpool
NAME      PROPERTY    VALUE    SOURCE
deadpool  recordsize  2M       local
[root@deadpool testing]# rm file.txt
rm: remove regular file 'file.txt'? y
[root@deadpool testing]#
```
The strace when recordsize is set to 2M and below:
```
execve("/usr/bin/rm", ["rm", "file.txt"], 0x7ffd54285608 /* 23 vars */) = 0
brk(NULL)                               = 0x5589459f5000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffd94a27bb0) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=164321, ...}) = 0
mmap(NULL, 164321, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fb2e7b8d000
close(3)                                = 0
openat(AT_FDCWD, "/lib64/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@@\2\0\0\0\0\0"..., 832) = 832
lseek(3, 792, SEEK_SET)                 = 792
read(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\2605Q\367\26\207U\27\277\347\342M\3541\24{"..., 68) = 68
fstat(3, {st_mode=S_IFREG|0755, st_size=5591560, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fb2e7b8b000
lseek(3, 792, SEEK_SET)                 = 792
read(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\2605Q\367\26\207U\27\277\347\342M\3541\24{"..., 68) = 68
lseek(3, 864, SEEK_SET)                 = 864
read(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32) = 32
mmap(NULL, 1857568, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fb2e79c5000
mprotect(0x7fb2e79e7000, 1679360, PROT_NONE) = 0
mmap(0x7fb2e79e7000, 1363968, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7fb2e79e7000
mmap(0x7fb2e7b34000, 311296, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x16f000) = 0x7fb2e7b34000
mmap(0x7fb2e7b81000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bb000) = 0x7fb2e7b81000
mmap(0x7fb2e7b87000, 14368, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fb2e7b87000
close(3)                                = 0
arch_prctl(ARCH_SET_FS, 0x7fb2e7b8c580) = 0
mprotect(0x7fb2e7b81000, 16384, PROT_READ) = 0
mprotect(0x558943f64000, 4096, PROT_READ) = 0
mprotect(0x7fb2e7be0000, 4096, PROT_READ) = 0
munmap(0x7fb2e7b8d000, 164321)          = 0
brk(NULL)                               = 0x5589459f5000
brk(0x558945a16000)                     = 0x558945a16000
brk(NULL)                               = 0x558945a16000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=217750512, ...}) = 0
mmap(NULL, 217750512, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fb2daa1b000
close(3)                                = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
newfstatat(AT_FDCWD, "file.txt", {st_mode=S_IFREG|0644, st_size=104857600, ...}, AT_SYMLINK_NOFOLLOW) = 0
geteuid()                               = 0
unlinkat(AT_FDCWD, "file.txt", 0)       = 0
lseek(0, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
close(0)                                = 0
close(1)                                = 0
close(2)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Thank you for all the hard work that you folks put in.

**Comment by loli10K** [comment]
Duplicate of https://github.com/zfsonlinux/zfs/issues/8778

**Comment by behlendorf** [comment]
Closing.  This issue has been fixed in master by commit ec4afd27f198d93a7bd32a05cb288708ba754ada and the fix will be applied to 0.8.1.  @JerkFace2017 you can safely cherry-pick ec4afd27f198d93a7bd32a05cb288708ba754ada if you can't wait for the point release.
